clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -639,56 +835,175 @@ public void assign(Collection<TopicPartition> partitions) {
+        if (subscriptions.subscribeFromPattern(topicsToSubscribe))
+            metadata.requestUpdateForNewTopics();
     }
 
     @Override
     public void subscribe(Pattern pattern) {
-        throw new KafkaException("method not implemented");
+        subscribe(pattern, new NoOpConsumerRebalanceListener());
     }
 
     @Override
     public void unsubscribe() {
-        throw new KafkaException("method not implemented");
+        fetchBuffer.retainAll(Collections.emptySet());
+        subscriptions.unsubscribe();
     }
 
     @Override
     @Deprecated
-    public ConsumerRecords<K, V> poll(long timeout) {
-        throw new KafkaException("method not implemented");
+    public ConsumerRecords<K, V> poll(final long timeoutMs) {
+        return poll(Duration.ofMillis(timeoutMs));
     }
 
     // Visible for testing
     WakeupTrigger wakeupTrigger() {
         return wakeupTrigger;
     }
 
-    private static <K, V> ClusterResourceListeners configureClusterResourceListeners(
-            final Deserializer<K> keyDeserializer,
-            final Deserializer<V> valueDeserializer,
-            final List<?>... candidateLists) {
-        ClusterResourceListeners clusterResourceListeners = new ClusterResourceListeners();
-        for (List<?> candidateList: candidateLists)
-            clusterResourceListeners.maybeAddAll(candidateList);
+    private Fetch<K, V> pollForFetches(Timer timer) {
+        long pollTimeout = timer.remainingMs();
+
+        // if data is available already, return it immediately
+        final Fetch<K, V> fetch = collectFetch();
+        if (!fetch.isEmpty()) {
+            return fetch;
+        }
+
+        // We do not want to be stuck blocking in poll if we are missing some positions
+        // since the offset lookup may be backing off after a failure
+
+        // NOTE: the use of cachedSubscriptionHasAllFetchPositions means we MUST call
+        // updateAssignmentMetadataIfNeeded before this method.
+        if (!cachedSubscriptionHasAllFetchPositions && pollTimeout > retryBackoffMs) {
+            pollTimeout = retryBackoffMs;
+        }
+
+        log.trace("Polling for fetches with timeout {}", pollTimeout);
+
+        Timer pollTimer = time.timer(pollTimeout);
 
-        clusterResourceListeners.maybeAdd(keyDeserializer);
-        clusterResourceListeners.maybeAdd(valueDeserializer);
-        return clusterResourceListeners;
+        // Wait a bit for some fetched data to arrive, as there may not be anything immediately available. Note the
+        // use of a shorter, dedicated "pollTimer" here which updates "timer" so that calling method (poll) will
+        // correctly handle the overall timeout.
+        try {
+            fetchBuffer.awaitNotEmpty(pollTimer);
+        } catch (InterruptException e) {
+            log.trace("Timeout during fetch", e);
+        } finally {
+            timer.update(pollTimer.currentTimeMs());
+        }
+
+        return collectFetch();
```

Comment:
Yes, we could update `pollForFetches()` to simply return `Fetch.empty()` at the end.

However, there's code that intentionally blocks (for a little bit) waiting for fetch data to arrive in the buffer:

```java
try {
    fetchBuffer.awaitNotEmpty(pollTimer);
} catch (InterruptException e) {
    log.trace("Timeout during fetch", e);
} finally {
    timer.update(pollTimer.currentTimeMs());
}

return collectFetch();
```

Also, each loop through `poll()` executes `updateAssignmentMetadataIfNeeded()` before checking the fetch buffer for any data. That method does a lot of work (network I/O) that we'd ideally skip if we already have the data.

Perhaps we could update `awaitNotEmpty()` with a return flag so that the tail end of `pollForFetches()` can look something like this:

```java
try {
    if (fetchBuffer.awaitNotEmpty(pollTimer))
        return collectFetch();
} catch (InterruptException e) {
    log.trace("Timeout during fetch", e);
} finally {
    timer.update(pollTimer.currentTimeMs());
}

return Fetch.empty();
```

Reply:
Thanks for the explanation, Kirk. We can just leave the code as it is then.